pkg/controller/service/service_controller_test.go

```diff
@@ -615,35 +616,61 @@ func TestProcessServiceCreateOrUpdate(t *testing.T) {
 
 }
 
-// TestConflictWhenProcessServiceCreateOrUpdate tests if processServiceCreateOrUpdate will
-// retry creating the load balancer when the update operation returns a conflict
-// error.
-func TestConflictWhenProcessServiceCreateOrUpdate(t *testing.T) {
-	svcName := "conflict-lb"
-	svc := newService(svcName, types.UID("123"), v1.ServiceTypeLoadBalancer)
-	controller, _, client := newController()
-	client.PrependReactor("update", "services", func(action core.Action) (bool, runtime.Object, error) {
-		update := action.(core.UpdateAction)
-		return true, update.GetObject(), apierrors.NewConflict(action.GetResource().GroupResource(), svcName, errors.New("Object changed"))
-	})
+// TestProcessServiceCreateOrUpdateK8sError tests processServiceCreateOrUpdate
+// with various kubernetes errors.
+func TestProcessServiceCreateOrUpdateK8sError(t *testing.T) {
+	svcName := "svc-k8s-err"
 
-	if err := controller.processServiceCreateOrUpdate(svc, svcName); err == nil {
-		t.Fatalf("controller.processServiceCreateOrUpdate() = nil, want error")
+	testCases := []struct {
+		desc      string
+		k8sErr    error
+		expectErr bool
+	}{
+		{
+			desc:      "conflict error",
+			k8sErr:    apierrors.NewConflict(schema.GroupResource{}, svcName, errors.New("Object conflict")),
+			expectErr: true,
+		},
+		{
+			desc:      "not found error",
+			k8sErr:    apierrors.NewNotFound(schema.GroupResource{}, svcName),
+			expectErr: false,
+		},
 	}
 
-	errMsg := "Error syncing load balancer"
-	if gotEvent := func() bool {
-		events := controller.eventRecorder.(*record.FakeRecorder).Events
-		for len(events) > 0 {
-			e := <-events
-			if strings.Contains(e, errMsg) {
-				return true
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			svc := newService(svcName, types.UID("123"), v1.ServiceTypeLoadBalancer)
+			controller, _, client := newController()
+			client.PrependReactor("patch", "services", func(action core.Action) (bool, runtime.Object, error) {
+				return true, nil, tc.k8sErr
+			})
+
+			err := controller.processServiceCreateOrUpdate(svc, svcName)
+			gotErr := err != nil
+			if gotErr != tc.expectErr {
```

Comment:
Can `expectErr` just be the actual error from `processServiceCreateOrUpdate`? That way we can just `reflect.DeepEqual` against the returned error.

Reply:
Yep that is better, updated.